### ISSUE TITLE
Refactor Block Tree RPC for Efficiency / Bandwidth Saving

### ISSUE THREAD

**Issue by rauljordan** [prysmaticlabs/prysm#2643]
Hi all,

Our current block tree RPC endpoint [here](https://github.com/prysmaticlabs/prysm/blob/632f6797cd41d212cdbb5f588835943e079da84b/beacon-chain/rpc/beacon_server.go#L285) returns a giant list of all the blocks from the last justified slot to the current, canonical slot. It is quite inefficient to fetch and to serve to clients. Additionally, we should specify request parameters to fetch the blocks within a range of slots.

Another useful feature would be to expose a separate endpoint for attestation targets and vote counts. One can compute the vote counts on a tree with the method as we currently do in our beacon_server.go in our RPC folder:

```go
	for _, kid := range fullBlockTree {
		if ctx.Err() != nil {
			return nil, ctx.Err()
		}
		votes, err := blockchain.VoteCount(kid, justifiedState, attestationTargets, bs.beaconDB)
		if err != nil {
			return nil, err
		}
		blockRoot, err := hashutil.HashBeaconBlock(kid)
		if err != nil {
			return nil, err
		}
		tree = append(tree, &pb.BlockTreeResponse_TreeNode{
			BlockRoot: blockRoot[:],
			Block:     kid,
			Votes:     uint64(votes),
		})
	}
```

**Comment by grasponcrypto** [comment]
``
Our current block tree RPC endpoint here returns a giant list of all the blocks from the last justified slot to the current, canonical slot. It is quite inefficient to fetch and to serve to clients. Additionally, we should specify request parameters to fetch the blocks within a range of slots.
``

Is the ask here to improve the efficiency of this call?

**Comment by katzdan** [comment]
I'd like to work on this one.

**Comment by terencechain** [comment]
before we start, I'd like to have a discussion on what 1.) request parameters and 2.) additional features we would implement with this issue.

can we give more directions here? @rauljordan

**Comment by katzdan** [comment]
I'll follow the discussion and ask some questions once I delve into the code. Any guidance you guys can provide before I start would be great. Thanks

**Comment by katzdan** [comment]
When can we talk about this? I would like to start working on it...thanks.

**Comment by terencechain** [comment]
@katzdan we can begin on adding a feature that allows request to query blocks given a range of slots, feel free to start on that

**Comment by katzdan** [comment]
Thanks

**Comment by prestonvanloon** [comment]
This needs an API design and design review before starting.

**Comment by katzdan** [comment]
How can we push this forward? 
I'm currently working on the Block filtering by slot range (pending fixing the test execution) but still waiting to see what else is in the scope of work for this issue.

**Comment by terencechain** [comment]
Will you be able to propose a few API designs on what you think would be useful for users to retrieving block tree?

**Comment by katzdan** [comment]
Sure.

**Comment by katzdan** [comment]
Hi Guys,

I'm trying to figure out which new APIs we can offer to 3rd party actors that want to query the beacon-chain service. Still learning , hope to have something within a few days. 
In any case - I can see the potential for a heavy load of data that's queries and then returned to the client. In many enterprise systems - the transaction related functionality is separated from reporting functionality to minimize the effect that reporting requests have on the transaction related operational capability. 

Since there is no separation between transactional and reporting services -it looks to me that querying the tree might have similar consequences here. Regardless of how clients can query the beacon-chain service - I would recommend that **any** function that's used to query the tree will support paging to reduce amount of data retrieved and returned to the client.
Your thoughts?

**Comment by katzdan** [comment]
I see this branch is merged so I'll create a new issue for adding new feature to the BlockTree rpc service.

**Comment by terencechain** [comment]
you can still use this issue, I merged the PR #2720 because what it's intend to do was done. We shouldn't implement a pr with multiple features, the goal is to keep it simple and easy to review

**Comment by terencechain** [comment]
Closing to favor #2739 for the new discussions

**Comment by prestonvanloon** [comment]
This issue has started a larger initiative to get a formal design on our entire read data API. Starting with this [product requirements document](https://docs.google.com/document/d/1LmlF0meuzx7LN8fSkuZWtGSZy1K0ZgyTIwL0zdR8Or0/edit), we're collecting use cases and feedback from users. Following that, we'll start on a formal design proposal for a proper API server schema.
